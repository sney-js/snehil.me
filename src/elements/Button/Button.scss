@import '../../styles/helpers/main';

$no-icon-hover: '&:not(.icon-secondary):not(.disabled):hover';

%button-style-light {
  --button-border-color: var(--color-primary);
  --button-background-hover: var(--color-primary);

  &.primary {
    --text-color: var(--color-light);
    --button-background: var(--color-primary);
    --button-background-hover: var(--color-primary-darker);

    #{$no-icon-hover} {
      --button-border-color: var(--button-background-hover);
    }
  }

  &.secondary {
    --button-background: transparent;
    --button-background-hover: var(--color-highlight);
  }
}

%button-style-dark {
  --button-border-color: var(--color-secondary);
  --button-background-hover: var(--color-secondary);

  &.primary {
    --button-background: var(--color-secondary);
    --text-color: var(--color-primary);
    --button-background-hover: var(--color-secondary-darker);

    #{$no-icon-hover} {
      --button-border-color: var(--button-background-hover);
    }
  }

  &.secondary {
    --text-color: inherit;
    --button-background-hover: var(--color-highlight);
  }
}

%button-vars {
  --text-color: var(--color-primary);
  --button-height: 42px;

  @extend %button-style-light;

  #{$themeDark},
  #{$themeDarkSelf} {
    @extend %button-style-dark;
  }

  &.disabled,
  &[disabled] {
    font-weight: normal;
    cursor: not-allowed;

    &.primary {
      --button-border-color: transparent;
      --button-background: var(--color-primary-darker);
      --text-color: var(--color-primary);
    }

    &.secondary {
      --button-border-color: var(--color-highlight-darker);
      --text-color: var(--color-highlight-darker);
    }
  }
}

.d-button {
  @extend %button-vars;
  @include setTypeStylesAll('button');

  position: relative;
  outline: none;
  text-align: center;
  cursor: pointer;
  margin: 0;
  box-sizing: border-box;
  min-height: var(--button-height);
  min-width: 152px;
  border-width: 2px;
  border-style: solid;
  transition-property: background-color;
  font-weight: 600;

  padding: 0px var(--spacing-gap);
  border-radius: var(--border-radius-button);
  border-color: var(--button-border-color);
  transition: var(--transition-ui);
  background-color: var(--button-background);
  color: var(--text-color);
  font-family: var(--font-primary);
  vertical-align: top;

  display: inline-flex;
  align-items: center;
  justify-content: space-evenly;

  &:focus {
    box-shadow: 0px 0px 0px 2px var(--color-highlight-secondary);
  }

  @media #{$screen-mobile} {
    width: 100%;
    display: block;
  }

  #{$no-icon-hover} {
    background-color: var(--button-background-hover);
  }

  // ------------ icon secondary --------------------

  &.icon-secondary {
    .d-button-icon {
      visibility: hidden;
      opacity: 0;
    }

    &.icon-secondary-show,
    &:hover,
    &:active,
    &:focus {
      padding-right: 48px;

      .d-button-icon {
        visibility: visible;
        opacity: 1;
        transform: translate(0%, -50%);
      }
    }

    .d-button-icon {
      line-height: 0;
      width: 24px;
      position: absolute;
      right: var(--spacing-gap);
      top: 50%;
      transform: translate(-10%, -50%);
      transition: var(--transition-ui);
    }
  }

  // ------------ icon primary --------------------

  &.icon-primary {
    text-align: left;
    min-width: var(--button-height);
    justify-content: space-between;

    @media #{$screen-mobile} {
      width: unset;
      display: inline-flex;
    }

    .d-button-icon {
      width: 24px;
      line-height: 0;
      position: absolute;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
      transition: var(--transition-ui);
    }
  }

  &.icon-primary-inline {
    .d-button-icon {
      position: static;
      transform: translate(0%, 0%);
      padding-left: var(--spacing-gap);
    }
  }
}
