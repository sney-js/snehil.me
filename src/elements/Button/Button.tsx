import React, { FC } from 'react';
import { makeClass } from 'utils/Helpers';
import Spinner from 'elements/Loaders';

export type ButtonProps = {
  /**
   * This will replace children if provided
   */
  title?: string;
  /**
   * One of 'primary' | 'secondary'
   */
  appearance?: 'primary' | 'secondary';
  /**
   * Displays icon on the right with text. Center without.
   */
  icon?: JSX.Element;
  /**
   * Add this prop for a loading indicator
   */
  isLoading?: boolean;
  /**
   * If the parent element should be a `<button>` or `<div>`
   */
  asDiv?: boolean;
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

/**
 * Use `Button` to highlight key info with a predefined status.
 */
const Button: FC<ButtonProps> = (props: ButtonProps) => {
  const { appearance, isLoading, asDiv, title, icon, ...rest } = props;

  const classes = makeClass([
    props.className,
    'd-button',
    appearance,
    props.disabled && 'disabled',
    icon && 'icon-primary',
    icon && (props.children || props.title) ? 'icon-primary-inline' : undefined,
    isLoading && !icon && !props.disabled && 'icon-secondary',
    !icon && isLoading && 'icon-secondary-show'
  ]);

  return React.createElement(
    asDiv ? 'div' : 'button',
    { className: classes, ...rest },
    <React.Fragment>
      {props.children || props.title}
      {isLoading && (
        <i className='d-button-icon'>
          <Spinner type='gg-spinner' />
        </i>
      )}
      {icon && <i className='d-button-icon'>{icon}</i>}
    </React.Fragment>
  );
};

Button.displayName = 'Button';
Button.defaultProps = {
  appearance: 'primary'
};
export default Button;
