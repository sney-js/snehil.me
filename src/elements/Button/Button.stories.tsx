import React from 'react';
import Button from './Button';

export default {
  title: 'Elements/Button',
  parameters: {
    componentSubtitle: 'Atom'
  },
  component: Button
};

export const basic = (args) => {
  return (
    <div>
      <Button title='Button' {...args} />
    </div>
  );
};
