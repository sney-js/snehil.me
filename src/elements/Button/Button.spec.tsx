import * as React from 'react';
import Button from './Button';
import { shallow } from 'enzyme';

describe('<Button />', () => {
  test('Basic', () => {
    const result = shallow(<Button title='no' />);
    expect(result.text()).toBe('no');

    const button = result.find('button');
    expect(button.type()).toBe('button');

    expect(result.html()).toMatchSnapshot();
  });

  test('Title as children', () => {
    const result = shallow(<Button>no</Button>);
    expect(result.text()).toBe('no');

    const button = result.find('button');
    expect(button.type()).toBe('button');
  });

  test('Appearance', () => {
    const result = shallow(
      <Button appearance='secondary' title='Secondary' />
    );

    const instance = result.find('button');
    expect(instance.props().className).toContain('secondary');

    expect(result).toMatchSnapshot();
  });
});
