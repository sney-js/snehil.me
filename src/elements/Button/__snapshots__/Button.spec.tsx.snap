// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Appearance 1`] = `ShallowWrapper {}`;

exports[`<Button /> Basic 1`] = `"<button class=\\"d-button primary\\">no</button>"`;
